src/gui/joytabwidget.cpp

```diff
@@ -1277,7 +1277,12 @@ void JoyTabWidget::changeCurrentSet(int index)
         oldSetButton->style()->polish(oldSetButton);
     }
 
-    m_joystick->setActiveSetNumber(index);
+    // Call only in reaction to a GUI button click to avoid race condition
+    if (QObject::sender()->metaObject()->className() == "QPushButton")
```

Comment:
Thank you for this fix! I believe that this unfortunately won't work because className() returns `const char *`? https://doc.qt.io/qt-6/qmetaobject.html#className

So, this equality is comparing the addresses and will always evaluate to false. It means that the UI buttons will no longer have any effect.

The approach that I took was to move `m_joystick->setActiveSetNumber(index);` into the UI handler functions (changeSetOne, changeSetTwo, etc) so that it would only be called from the UI and not from the signal, but I will leave it up to you how you want to implement this :)

Reply:
> So, this equality is comparing the addresses and will always evaluate to false

You are right.

> The approach that I took was to move m_joystick->setActiveSetNumber(index); into the UI handler functions (changeSetOne, changeSetTwo, etc) so that it would only be called from the UI and not from the signal

I don't like these kinds of repetitions and scattering a logic doing one thing across multiple places in code. 
I think that removing the line   https://github.com/AntiMicroX/antimicrox/blob/2f7b66be195d8b3131bce9a250149ff3e7a62db9/src/inputdevice.cpp#L1135
would be the most elegant, because we would just get rid of this unnecessary call.